### PR TITLE
Simplify tracked state at each summarizerNode

### PR DESCRIPTION
**Pull request by pragya91**
This PR removes some redundant tracked states that are maintained by each summarizerNode. Following are the list of changes:
- Disable incremental summaries and perform complete summarization instead (the first time), for the cases where base snapshot was created in an older document where isolated channels were disabled (i.e. `.channels` concept was not there in a node path). [`containerRuntime.ts`]

- `SummaryNode` is removed. SummaryNode was tracking base, additional etc. path related information for each node, which was convoluting the code too much. It has been simplified by always providing the path information to the summarizerNode as a `_summaryHandleId`. Every time a child node it created, it's path is stored as _summaryHandleId and thus removes the need for `SummaryNode`. 
[`summarizerNode.ts`, `summarizerNodeUtils.ts`].

- `latestSummary:SummaryNode` is replaced with just `_lastSummaryreferenceSequenceNumber`.

- `PendingSummaryInfo` has been introduced to solve for the other half of the logic that was fulfilled previously by `SummaryNode` i.e. maintaining the state of pending summaries submitted so far. Instead of saving path and referenceSequenceNumber both, we now just maintain `referenceSequenceNumber`. Saves memory footprint, and simplifies code. [`summarizerNode.ts`, `summarizerNodeUtils.ts`]

- The above changes also reflected in`summarizerNodeWithGC.ts`.

### CODE REVIEW

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 1]
```diff
@@ -1649,15 +1649,21 @@ export class ContainerRuntime
 		});
 
 		const loadedFromSequenceNumber = this.deltaManager.initialSequenceNumber;
+		// If the base snapshot was generated when isolated channels were disabled, set the summary reference
+		// sequence to undefined so that this snapshot will not be used for incremental summaries. This is for
+		// back-compat and will rarely happen so its okay to re-summarize everything in the first summary.
+		const summaryReferenceSequenceNumber =
+			baseSnapshot === undefined || metadata?.disableIsolatedChannels === true
```
Comment (tyler-cai-microsoft) [inline packages/runtime/container-runtime/src/containerRuntime.ts:1656]: Where does this extra `metadata?.disableIsolatedChannels === true` come from?

File: packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNode.ts [hunk 2]
```diff
@@ -579,26 +550,26 @@ export class SummarizerNode implements IRootSummarizerNode {
...
 		let changeSequenceNumber: number;
 
-		const parentLatestSummary = this._latestSummary;
+		const parentLastSummaryReferenceSequenceNumber = this._lastSummaryReferenceSequenceNumber;
 		switch (createParam.type) {
 			case CreateSummarizerNodeSource.FromAttach: {
 				if (
-					parentLatestSummary !== undefined &&
-					createParam.sequenceNumber <= parentLatestSummary.referenceSequenceNumber
+					this._lastSummaryReferenceSequenceNumber !== undefined &&
```
Comment (tyler-cai-microsoft) [inline packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNode.ts:560]: it seems like this should be `parentLastSummaryReferenceSequenceNumber` though there is no real difference technically

File: packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNode.ts [hunk 3]
```diff
@@ -608,11 +579,16 @@ export class SummarizerNode implements IRootSummarizerNode {
...
+			EscapedPath.create(id),
+			channelsTreeName,
+		);
 
 		return {
-			latestSummary,
 			changeSequenceNumber,
 			telemetryNodeId: childTelemetryNodeId,
+			summaryHandleId: childSummaryHandleId,
+			lastSummaryreferenceSequenceNumber: childLastSummaryReferenceSequenceNumber,
```
Comment (tyler-cai-microsoft) [inline packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNode.ts:591]: Capitalize the R in reference?

File: packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNodeUtils.ts [hunk 4]
```diff
@@ -70,100 +69,33 @@ export class EscapedPath {
...
+	public concatWith(path: EscapedPath, concatWith: string): EscapedPath {
+		return new EscapedPath(`${this.path}/${encodeURIComponent(concatWith)}/${path.path}`);
+	}
 	public concat(path: EscapedPath): EscapedPath {
 		return new EscapedPath(`${this.path}/${path.path}`);
 	}
 }
-
-/** Information about a summary relevant to a specific node in the tree */
-export class SummaryNode {
```
Comment (tyler-cai-microsoft) [inline packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNodeUtils.ts:91]: I feel like this can be deleted in a separate PR

File: packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNodeWithGc.ts [hunk 5]
```diff
@@ -422,22 +413,18 @@ export class SummarizerNodeWithGC extends SummarizerNode implements IRootSummari
 		}
 
 		// In case we have pending summaries on the parent, let's initialize it on the child.
-		if (child.latestSummary !== undefined) {
-			for (const [key, value] of this.pendingSummaries.entries()) {
-				const summaryNodeWithGC = value as SummaryNodeWithGC;
-				if (summaryNodeWithGC.serializedUsedRoutes !== undefined) {
+		if (child.referenceSequenceNumber !== undefined) {
+			for (const [key, pendingSummary] of this.pendingSummaries.entries()) {
+				const pendingSummaryeWithGC = pendingSummary as PendingSummaryInfoWithGC;
```
Comment (tyler-cai-microsoft) [inline packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNodeWithGc.ts:418]: I don't know if this is possible, is there a way to remove this cast? Also, we don't need to do that in this PR

File: packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNode.ts [hunk 6]
```diff
@@ -63,13 +63,27 @@ export class SummarizerNode implements IRootSummarizerNode {
...
+	/** Key value pair of summaries submitted by this client which are not yet acked. Key is the proposalHandle and value is the summary op's referece sequence number */
+	protected readonly pendingSummaries = new Map<string, PendingSummaryInfo>();
 	protected wipReferenceSequenceNumber: number | undefined;
-	private wipLocalPaths: { localPath: EscapedPath; additionalPath?: EscapedPath } | undefined;
+	/**
+	 * True if the current node was summarized during the current summary process
+	 * This flag is used to identify scenarios where a node may not have been summarized during the summary
+	 * because of application code creating data structures for data stores which were already summarized.
+	 */
+	private wipSummary: boolean = false;
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNode.ts:86]: Consider renaming this to something that conveys whether this node was summarizer, like `wipSummarizeCalled` or `wipWasSummarized`.

File: packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNode.ts [hunk 7]
```diff
@@ -322,31 +327,14 @@ export class SummarizerNode implements IRootSummarizerNode {
...
-				// by its last known good state in the actual summary tree.
-				// If parent fails or is reused, the child summarize is not called so
-				// it did not get a chance to change its paths.
-				// In this case, essentially only propagate the new summary ref seq num.
-				localPathsToUse = {
-					localPath: latestSummary.localPath,
-					additionalPath: latestSummary.additionalPath,
-				};
-			} else {
+			const latestSummary = this._lastSummaryReferenceSequenceNumber;
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNode.ts:336]: nit: You don't need to assign it to a local variable.

File: packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNode.ts [hunk 8]
```diff
@@ -461,48 +443,40 @@ export class SummarizerNode implements IRootSummarizerNode {
...
-	}
-
 	public updateBaseSummaryState(snapshot: ISnapshotTree) {
-		// Check base summary to see if it has any additional path parts
-		// separating child SummarizerNodes. Checks for .channels subtrees.
-		const { childrenPathPart } = parseSummaryForSubtrees(snapshot);
-		if (childrenPathPart !== undefined && this._latestSummary !== undefined) {
-			this._latestSummary.additionalPath = EscapedPath.create(childrenPathPart);
-		}
+		// Function deprecated. Empty declaration is kept around to compat failures.
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNode.ts:479]: Add a @deprecated tag to the function doc comments and add explanation there.

File: packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNode.ts [hunk 9]
```diff
@@ -632,20 +608,14 @@ export class SummarizerNode implements IRootSummarizerNode {
...
-				const newLatestSummaryNode = new SummaryNode({
-					referenceSequenceNumber: value.referenceSequenceNumber,
-					basePath: child._latestSummary.basePath,
-					localPath: child._latestSummary.localPath,
-				});
-
-				child.addPendingSummary(key, newLatestSummaryNode);
-			}
+		if (child._lastSummaryReferenceSequenceNumber !== undefined) {
+			this.pendingSummaries.forEach((summary, proposedHandle) => {
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNode.ts:612]: nit:
```suggestion
			this.pendingSummaries.forEach((summaryInfo, proposedHandle) => {
```

File: packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNode.ts [hunk 10]
```diff
@@ -632,20 +608,14 @@ export class SummarizerNode implements IRootSummarizerNode {
...
+				child.addPendingSummary(proposedHandle, summary);
+			});
 		}
 	}
 
-	protected addPendingSummary(key: string, summary: SummaryNode) {
+	protected addPendingSummary(key: string, summary: PendingSummaryInfo) {
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNode.ts:618]: nit:
```suggestion
	protected addPendingSummary(key: string, summaryInfo: PendingSummaryInfo) {
```

File: packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNodeWithGc.ts [hunk 11]
```diff
@@ -317,7 +307,7 @@ export class SummarizerNodeWithGC extends SummarizerNode implements IRootSummari
 			const summaryNode = this.pendingSummaries.get(proposalHandle);
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNodeWithGc.ts:307]:
```suggestion
			const pendingSummaryInfo = this.pendingSummaries.get(proposalHandle);
```

File: packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNodeWithGc.ts [hunk 12]
```diff
@@ -422,22 +413,18 @@ export class SummarizerNodeWithGC extends SummarizerNode implements IRootSummari
 		}
 
 		// In case we have pending summaries on the parent, let's initialize it on the child.
-		if (child.latestSummary !== undefined) {
-			for (const [key, value] of this.pendingSummaries.entries()) {
-				const summaryNodeWithGC = value as SummaryNodeWithGC;
-				if (summaryNodeWithGC.serializedUsedRoutes !== undefined) {
+		if (child.referenceSequenceNumber !== undefined) {
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNodeWithGc.ts:416]: `referenceSequenceNumber` will never be undefined.

File: packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNodeWithGc.ts [hunk 13]
```diff
@@ -422,22 +413,18 @@ export class SummarizerNodeWithGC extends SummarizerNode implements IRootSummari
...
 					const newSerializedRoutes = childNodeUsedRoutes.get(id) ?? [""];
-					const newLatestSummaryNode = new SummaryNodeWithGC(
-						JSON.stringify(newSerializedRoutes),
-						{
-							referenceSequenceNumber: value.referenceSequenceNumber,
-							basePath: child.latestSummary.basePath,
-							localPath: child.latestSummary.localPath,
-						},
-					);
+					const newLatestSummaryNode = {
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNodeWithGc.ts:424]:
```suggestion
					const childPendingSummaryInfo = {
```

File: packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNodeWithGc.ts (hunk 5 above)
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNodeWithGc.ts:418]:
```suggestion
				const pendingSummaryWithGC = pendingSummary as PendingSummaryInfoWithGC;
```

**Comment by agarwal-navin** [comment]
Excellent! Very excited for this change and the complexity / code it removes!

File: packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNodeUtils.ts [hunk 14]
```diff
@@ -70,100 +69,33 @@ export class EscapedPath {
...
-			// Non null asserting here since we are iterating over pathParts
-			// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
-			ret = ret.concat(EscapedPath.create(pathParts[i]!));
-		}
-		return ret;
-	}
 	public toString(): string {
 		return this.path;
 	}
+	public concatWith(path: EscapedPath, concatWith: string): EscapedPath {
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNodeUtils.ts:75]: It's not obvious what this function does based on the param names.
I would rename the params to something like `path1: string` and `path2: EscapedPath` so it's clear what order they will be concatenated in.
Also, maybe you could just accept strings and do the `EscapedPath.create` part here to keep it consistent.

Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNodeUtils.ts:75]: nit: Can you add a comment to this function to make it clear what it is doing.

File: packages/runtime/runtime-definitions/src/summary.ts [hunk 15]
```diff
@@ -209,6 +209,9 @@ export interface ISummarizerNode {
 	 * that it might have. For example: if datastore "a" contains dds "b", but the
 	 * path is "/a/.channels/b", then the additional path part is ".channels".
 	 * @param snapshot - the base summary to parse
+	 *
+	 * @deprecated - The code now always assumes that all summary nodes have .channels
```
Comment (Josmithr) [inline packages/runtime/runtime-definitions/src/summary.ts:213]: Nit: `@deprecated` comments don't want a `-` between the tag and the body. The only TSDoc tags that want that are `@param` blocks.
```suggestion
	 * @deprecated The code now always assumes that all summary nodes have .channels
```

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 1 above)
Comment (pragya91) [reply packages/runtime/container-runtime/src/containerRuntime.ts:1656]: From ISummaryRuntimeOptions.

File: packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNodeUtils.ts [hunk 16]
```diff
@@ -70,100 +69,33 @@ export class EscapedPath {
...
+		return new EscapedPath(`${this.path}/${encodeURIComponent(concatWith)}/${path.path}`);
+	}
 	public concat(path: EscapedPath): EscapedPath {
```
Comment (rajatch-ff) [inline packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNodeUtils.ts:78]: How's this different from concatWith above? can we highlight the difference in the signature?

File: packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNodeWithGc.ts [hunk 17]
```diff
@@ -37,18 +37,8 @@ export interface IRootSummarizerNodeWithGC
 		ISummarizerNodeRootContract {}
 
 // Extend SummaryNode to add used routes tracking to it.
```
Comment (rajatch-ff) [inline packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNodeWithGc.ts:39]: update comment

File: packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNodeWithGc.ts [hunk 18]
```diff
@@ -268,28 +260,26 @@ export class SummarizerNodeWithGC extends SummarizerNode implements IRootSummari
...
+		super.completeSummaryCore(proposalHandle, parentSkipRecursion);
 
 		// If GC is disabled, skip setting pending summary with GC state.
 		if (!this.gcDisabled) {
-			const summaryNode = this.pendingSummaries.get(proposalHandle);
-			if (summaryNode !== undefined) {
-				const summaryNodeWithGC = new SummaryNodeWithGC(wipSerializedUsedRoutes, summaryNode);
+			const pendingSummaryInfo = this.pendingSummaries.get(proposalHandle);
+			if (pendingSummaryInfo !== undefined) {
+				const summaryNodeWithGC = {
```
Comment (rajatch-ff) [inline packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNodeWithGc.ts:279]: rename var `summaryNodeWithGC` to something like `pendingSummaryInfoWithGC`

File: packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNodeWithGc.ts [hunk 19]
```diff
@@ -317,7 +307,7 @@ export class SummarizerNodeWithGC extends SummarizerNode implements IRootSummari
 			const summaryNode = this.pendingSummaries.get(proposalHandle);
 			if (summaryNode !== undefined) {
 				// If a pending summary exists, it must have used routes since GC is enabled.
-				const summaryNodeWithGC = summaryNode as SummaryNodeWithGC;
+				const summaryNodeWithGC = summaryNode as PendingSummaryInfoWithGC;
```
Comment (rajatch-ff) [inline packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNodeWithGc.ts:310]: rename var `summaryNodeWithGC` to something like `pendingSummaryInfoWithGC`

File: packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNodeWithGc.ts [hunk 20]
```diff
@@ -530,10 +517,9 @@ export const createRootSummarizerNodeWithGC = (
 		logger,
 		summarizeInternalFn,
 		config,
+		EscapedPath.create("") /* summaryHandleId */,
 		changeSequenceNumber,
-		referenceSequenceNumber === undefined
-			? undefined
-			: SummaryNode.createForRoot(referenceSequenceNumber),
+		referenceSequenceNumber,
```
Comment (rajatch-ff) [inline packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNodeWithGc.ts:522]: Can't this be undefined anymore? if yes, update the param-header

Comment (pragya91) [reply packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNodeWithGc.ts:520]: Can be undefined here

File: packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNodeUtils.ts (hunk 4 above)
Comment (pragya91) [reply packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNodeUtils.ts:91]: Related. Not splitting in a separate PR.

File: packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNode.ts (hunk 8 above)
Comment (pragya91) [reply packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNode.ts:478]: Done in the interface.

File: packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNodeUtils.ts (hunk 14 above)
Comment (pragya91) [reply packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNodeUtils.ts]: updated funciton name and signature

File: packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNodeUtils.ts (hunk 16 above)
Comment (pragya91) [reply packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNodeUtils.ts]: removed unused function concat

File: packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNode.ts [hunk 21]
```diff
@@ -63,13 +62,27 @@ export class SummarizerNode implements IRootSummarizerNode {
...
+	 * returns the handle of the last successful summary of this summarizerNode in string format
+	 * (this getter is primarily only used in the test code)
+	 */
+	public get summaryHandleId(): string {
+		return this._summaryHandleId.toString();
 	}
 
 	protected readonly children = new Map<string, SummarizerNode>();
-	protected readonly pendingSummaries = new Map<string, SummaryNode>();
+	/** Key value pair of summaries submitted by this client which are not yet acked. Key is the proposalHandle and value is the summary op's referece sequence number */
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNode.ts:77]: nit: formatting
```suggestion
	/**
	 *Key value pair of summaries submitted by this client which are not yet acked. Key is the proposalHandle and value is the summary op's referece sequence number\
	 */
```

File: packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNode.ts [hunk 22]
```diff
@@ -63,13 +62,27 @@ export class SummarizerNode implements IRootSummarizerNode {
...
+	protected readonly pendingSummaries = new Map<string, PendingSummaryInfo>();
 	protected wipReferenceSequenceNumber: number | undefined;
-	private wipLocalPaths: { localPath: EscapedPath; additionalPath?: EscapedPath } | undefined;
+	/**
+	 * True if the current node was summarized during the current summary process
+	 * This flag is used to identify scenarios where a node may not have been summarized during the summary
+	 * because of application code creating data structures for data stores which were already summarized.
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNode.ts:83]: `because of application code creating data structures for data stores which were already summarized`

Not sure what this means. Maybe this should say something like "to identify scenarios where summarize was not called on a node. For example, this node was created after its parent was already summarized due to out-of-order realization via application code".

File: packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNode.ts [hunk 23]
```diff
@@ -165,26 +180,24 @@ export class SummarizerNode implements IRootSummarizerNode {
 		trackState: boolean = true,
 		telemetryContext?: ITelemetryContext,
 	): Promise<ISummarizeResult> {
+		// Set to wipSummarizeCalled true to represent that current node was included in the summary process.
+		this.wipSummarizeCalled = true;
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNode.ts:184]: don't set it if `trackState` is faise. This is changing behavior.

File: packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNode.ts [hunk 24]
```diff
@@ -165,26 +180,24 @@ export class SummarizerNode implements IRootSummarizerNode {
...
-				};
+			if (this._lastSummaryReferenceSequenceNumber !== undefined) {
 				this.wipSkipRecursion = true;
 				const stats = mergeStats();
 				stats.handleNodeCount++;
 				return {
 					summary: {
 						type: SummaryType.Handle,
-						handle: latestSummary.fullPath.path,
+						handle: this._summaryHandleId.toString(),
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNode.ts:200]:
```suggestion
						handle: this.summaryHandleId,
```

File: packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNode.ts [hunk 25]
```diff
@@ -199,12 +212,12 @@ export class SummarizerNode implements IRootSummarizerNode {
...
-				this._latestSummary !== undefined
+				this._lastSummaryReferenceSequenceNumber !== undefined
 					? {
 							summarySequenceNumber: this.wipReferenceSequenceNumber,
-							latestSummarySequenceNumber: this._latestSummary.referenceSequenceNumber,
-							// TODO: remove summaryPath
-							summaryPath: this._latestSummary.fullPath.path,
+							latestSummarySequenceNumber: this._lastSummaryReferenceSequenceNumber,
+							// TODO: remove summaryPath.
+							summaryPath: this._summaryHandleId.toString(),
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNode.ts:220]:
```suggestion
							summaryPath: this.summaryHandleId,
```

File: packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNode.ts [hunk 26]
```diff
@@ -281,7 +289,7 @@ export class SummarizerNode implements IRootSummarizerNode {
 		// If the parent node skipped recursion, it did not call summarize on this node. So, summarize was not missed
 		// but was intentionally not called.
 		// Otherwise, summarize should have been called on this node and wipLocalPaths must be set.
-		if (parentSkipRecursion || this.wipLocalPaths !== undefined) {
+		if (parentSkipRecursion || this.wipSummarizeCalled) {
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNode.ts:292]: Update comment to say `wipSummarizeCalled` instead of `wipLocalPaths`

File: packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNode.ts [hunk 27]
```diff
@@ -398,7 +373,8 @@ export class SummarizerNode implements IRootSummarizerNode {
 	 * Refreshes the latest summary tracked by this node. If we have a pending summary for the given proposal handle,
 	 * it becomes the latest summary. If the current summary is already ahead, we skip the update.
 	 * If the current summary is behind, then we do not refresh.
-	 *
+	 * @param proposalHandle - Handle from the summary ack op
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNode.ts:376]:
```suggestion
	 * @param proposalHandle - Handle of the generated / uploaded summary.
```
Note: The handle in summary ack will be different from propsal handle in case of 2 commit summaries

File: packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNode.ts [hunk 28]
```diff
@@ -398,7 +373,8 @@ export class SummarizerNode implements IRootSummarizerNode {
...
+	 * @param summaryRefSeq - Reference sequence of the summary ack op
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNode.ts:377]:
```suggestion
	 * @param summaryRefSeq - Reference sequence of the acked summary
```

File: packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNode.ts [hunk 29]
```diff
@@ -579,26 +548,26 @@ export class SummarizerNode implements IRootSummarizerNode {
...
 		let changeSequenceNumber: number;
 
-		const parentLatestSummary = this._latestSummary;
+		const parentLastSummaryReferenceSequenceNumber = this._lastSummaryReferenceSequenceNumber;
 		switch (createParam.type) {
 			case CreateSummarizerNodeSource.FromAttach: {
 				if (
-					parentLatestSummary !== undefined &&
-					createParam.sequenceNumber <= parentLatestSummary.referenceSequenceNumber
+					this._lastSummaryReferenceSequenceNumber !== undefined &&
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNode.ts:558]: nit: I suggest using either `parentLastSummaryReferenceSequenceNumber` throughout this function of not creating that variable in the first place. Right now, it's using a combination of `parentLastSummaryReferenceSequenceNumber` and `this._lastSummaryReferenceSequenceNumber` which seems weird.

File: packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNode.ts (hunk 24 above)
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNode.ts:200]: The ids of data stores and DDSes can now be special characters which will get encoded. Please make sure that the encoding happens correctly such that it matches with the ids in the summary tree.
For instance, say a DDS id is `/@/&`. The summary tree will have a sub-tree for data store with id `@` (or the encoded format) and a sub-tree for a DDS under it with id `&` (or the encoded format). The handle is here should match the id for that tree.
It looks like you kept the logic the same, but it's hard to tell from the code and we had a bug like that already in the driver code.

File: packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNodeUtils.ts [hunk 30]
```diff
@@ -64,104 +64,46 @@ export interface ISummarizerNodeRootContract {
...
-			localPath: EscapedPath.create(id),
-		});
+	public createChildPath(childNodePath: EscapedPath): EscapedPath {
+		return new EscapedPath(
+			`${this.path}/${encodeURIComponent(channelsTreeName)}/${childNodePath.path}`,
+		);
 	}
 }
+export interface PendingSummaryInfo {
+	/** The sequence number up until which the summary was created. */
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNodeUtils.ts:91]: nit
```suggestion
	/** The sequence number at which the summary was created. */
```

File: packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNodeUtils.ts [hunk 31]
```diff
@@ -64,104 +64,46 @@ export interface ISummarizerNodeRootContract {
...
 export interface ICreateChildDetails {
-	/** Latest summary from server node data */
-	latestSummary: SummaryNode | undefined;
 	/** Sequence number of latest known change to the node */
 	changeSequenceNumber: number;
 	/** A unique id of this child to be logged when sending telemetry. */
 	telemetryNodeId: string;
+	/** Summary handle for child node */
+	summaryHandleId: EscapedPath;
+	/** last reference sequence number seen when last successful summary was created */
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNodeUtils.ts:105]: This doesn't seem quite right. It is more like the the reference sequence number of the last successful summary.

File: packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNodeWithGc.ts [hunk 32]
```diff
@@ -103,9 +93,10 @@ export class SummarizerNodeWithGC extends SummarizerNode implements IRootSummari
 		logger: ITelemetryBaseLogger,
 		summarizeInternalFn: SummarizeInternalFn,
 		config: ISummarizerNodeConfigWithGC,
+		_summaryHandleId: EscapedPath,
 		changeSequenceNumber: number,
-		/** Undefined means created without summary */
-		latestSummary?: SummaryNode,
+		/** Summary reference sequence number, i.e. last sequence number seen when it was created */
+		_lastSummaryReferenceSequenceNumber?: number,
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNodeWithGc.ts:99]: nit: don't use underscore since this class doesn't do anything with these two properties
```suggestion
		lastSummaryReferenceSequenceNumber?: number,
```

File: packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNode.ts (hunk 25 above)
Comment (pragya91) [reply packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNode.ts:220]: @agarwal-navin  --- the type of the summaryPath is a string in `IExperimentalIncrementalSummaryContext`.

File: packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNode.ts (hunk 24 above)
Comment (pragya91) [reply packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNode.ts:200]: @agarwal-navin --- the type of the handle is a string in `ISummaryHandle`.

Comment (pragya91) [reply packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNode.ts:204]: Added a new test file - `summarizerNodeUtils.spec.ts`

Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNode.ts:204]: The type of `summaryHandle` in this class is also string. it does `this._summaryHandleId.toString()` for you so there is no need to do it here again.

File: packages/runtime/container-runtime/src/test/summaryNodeUtils.spec.ts [hunk 33]
```diff
@@ -0,0 +1,39 @@
...
+
+	it("Validate basic path append behavior with special characters in path", () => {
+
+		const parentPath = EscapedPath.create("@")
+		assert.equal(parentPath.toString(), "%40");
+
+		const childPath = EscapedPath.create("&")
+		assert.equal(childPath.toString(), "%26");
+
+		const completePath = parentPath.createChildPath(childPath);
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/test/summaryNodeUtils.spec.ts:36]: For completeness, you also need to validate that the summary generated with these ids will have the encoded path for these ids in the path. Actually, I am not sure if that needs to be the case. The driver does some encoding of the summary before upload I believe. I would check with Vlad / Jatin to be sure.

**Review by agarwal-navin — APPROVED** [review]

File: packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNodeUtils.ts [hunk 34]
```diff
@@ -64,104 +64,46 @@ export interface ISummarizerNodeRootContract {
...
 export interface ICreateChildDetails {
-	/** Latest summary from server node data */
-	latestSummary: SummaryNode | undefined;
 	/** Sequence number of latest known change to the node */
 	changeSequenceNumber: number;
 	/** A unique id of this child to be logged when sending telemetry. */
 	telemetryNodeId: string;
+	/** Summary handle for child node */
+	summaryHandleId: EscapedPath;
+	/** the reference sequence number of the last successful summary. */
```
Comment (Josmithr) [inline packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNodeUtils.ts:105]: Nit
```suggestion
	/** The reference sequence number of the last successful summary. */
```

**Review by Josmithr — APPROVED** [review]
API changes look good :)
